### ISSUE TITLE
baseUrl is incorrectly generated in child route when id parameter contains whitespace or other special characters

### ISSUE THREAD

**Issue by mzhang444** [aurelia/router#486]
**I'm submitting a bug report**

* **Library Version:**
1.2.1

**Please tell us about your environment:**
* **Operating System:**
OSX 10.11.13

* **Node Version:**
6.9.1

* **NPM Version:**
3.10.8

* **Browser:**
Chrome 56.0.29

* **Language:**
ESNext

**Current behavior:**
When "id" parameter is nested child route contains whitespace or other special characters that encodeURI will encode, the baseUrl in the child route are not calculated correctly.

**Expected/desired behavior:**

I have an application with following nested child routes defined:
users/:id
   - access-ctrl
       - accounts
       - accounts/:id
       - others
When account details is displayed, I have links to go back to user details, access control, accounts list page.

When account id is "abc", all the links are created correctly:
Account details page: http://localhost:9000/#/users/user1/access-ctrl/accounts/abc
User details page: http://localhost:9000/#/users/user1
Access control page: http://localhost:9000/#/users/user1/access-ctrl
Account list page: http://localhost:9000/#/users/user1/access-ctrl/accounts

When account is is "test 1", baseUrl becomes empty string for User details and account list page:
Account details page: http://localhost:9000/#/users/user1/access-ctrl/accounts/test 1
User details page: http://localhost:9000/#
Access control page: http://localhost:9000/#/access-ctrl
Account list page: http://localhost:9000/#/accounts

If I encode the whitespace to %20, then User details and and first level child route Access Control links are generated correctly, but not the nested Account list child route:
Account details page: http://localhost:9000/#/users/user1/access-ctrl/accounts/test%201
User details page: http://localhost:9000/#/users/user1
Access control page: http://localhost:9000/#/users/user1/access-ctrl
Account list page: http://localhost:9000/#/users/user1/accounts

* **What is the expected behavior?**
Whitespace and other special characters in "id" parameter should be handled properly. Users should not need to encode special characters. If special characters need to be encoded by the user, then the encoded value should be passed into the nested route instead of decoded value. In my case, even if I did encode the account id, it only works at level 1 child route. By the time it reaches level 2 child route, the id value got decoded and no longer works.

Looking at the source code, the issue is caused by this method. In the method, path is encode, while fragment is not encoded, therefore, this method can return empty string when path has special characters.

```
NavigationInstruction.prototype.getBaseUrl = function getBaseUrl() {
      var _this = this;
      var fragment = this.fragment;

      if (fragment === '') {
        var nonEmptyRoute = this.router.routes.find(function (route) {
          return route.name === _this.config.name && route.route !== '';
        });
        if (nonEmptyRoute) {
          fragment = nonEmptyRoute.route;
        }
      }

      if (!this.params) {
        return fragment;
      }

      var wildcardName = this.getWildCardName();
      var path = this.params[wildcardName] || '';

      if (!path) {
        return fragment;
      }

      path = encodeURI(path);
      return fragment.substr(0, fragment.lastIndexOf(path));
```

* **What is the motivation / use case for changing the behavior?**

**Comment by PWKad** [comment]
@mzhang444 Can you create a failing spec(s) for this?  If it is simply to encode the fragment that should be an easy PR.

**Comment by InfiniteLukeOne** [comment]
I think this is the same problem as aurelia/templating-router#60 and should be fixed by #501
